paint/support/procedural/src/construct_runtime/parse.rs

```diff
@@ -0,0 +1,354 @@
+    syn::custom_keyword!(Block);
+    syn::custom_keyword!(NodeBlock);
+    syn::custom_keyword!(UncheckedExtrinsic);
+}
+
+#[derive(Debug)]
+pub struct RuntimeDefinition {
+    pub visibility_token: Token![pub],
+    pub enum_token: Token![enum],
+    pub name: Ident,
+    pub where_section: WhereSection,
+    pub modules: ext::Braces<ext::Punctuated<ModuleDeclaration, Token![,]>>,
+}
+
+impl Parse for RuntimeDefinition {
+    fn parse(input: ParseStream) -> Result<Self> {
+        Ok(Self {
+            visibility_token: input.parse()?,
+            enum_token: input.parse()?,
+            name: input.parse()?,
+            where_section: input.parse()?,
+            modules: input.parse()?,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct WhereSection {
+    pub block: syn::TypePath,
+    pub node_block: syn::TypePath,
+    pub unchecked_extrinsic: syn::TypePath,
+}
+
+impl Parse for WhereSection {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let _: Token![where] = input.parse()?;
+        let mut seen_keys = HashSet::new();
+        let mut definitions = HashMap::new();
+        while !input.peek(token::Brace) {
+            let WhereDefinition { kind, value } = WhereDefinition::parse(input, &mut seen_keys)?;
+            definitions.insert(kind, value);
+            if !input.peek(Token![,]) {
+                if !input.peek(token::Brace) {
+                    return Err(input.error("Expected `,` or `{`"));
+                }
+                break;
+            }
+            let _: Token![,] = input.parse()?;
+        }
+        let expected_seen_keys: HashSet<_> = vec![
+            WhereKind::Block,
+            WhereKind::NodeBlock,
+            WhereKind::UncheckedExtrinsic,
+        ]
+        .into_iter()
+        .collect();
+        let diff: Vec<_> = expected_seen_keys.difference(&seen_keys).collect();
+        if diff.len() > 0 {
+            let msg = format!(
+                "Missing associated type for `{:?}`. Add `{:?}` = ... to where section.",
+                diff[0], diff[0]
+            );
+            return Err(input.error(msg));
+        }
+        Ok(Self {
+            block: definitions
+                .remove(&WhereKind::Block)
+                .expect("Definitions is guaranteed to have this key; qed"),
+            node_block: definitions
+                .remove(&WhereKind::NodeBlock)
+                .expect("Definitions is guaranteed to have this key; qed"),
+            unchecked_extrinsic: definitions
+                .remove(&WhereKind::UncheckedExtrinsic)
+                .expect("Definitions is guaranteed to have this key; qed"),
+        })
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash)]
+pub enum WhereKind {
+    Block,
+    NodeBlock,
+    UncheckedExtrinsic,
+}
+
+impl WhereKind {
+    pub fn parse(input: ParseStream, seen_keys: &mut HashSet<WhereKind>) -> Result<Self> {
+        let lookahead = input.lookahead1();
+        if lookahead.peek(keyword::Block) {
```

Comment:
I made it to avoid storing keyword tokens inside the `WhereKind` enum (storing keyword tokens is required for meaningful error messages for duplicates). This would introduce too much pain with manually implementing Hash, PartialEq, etc. for `WhereKind`.

However, I found a better approach that looks good from every perspective.